docs/okeps/okep-5377-extend-udn-to-support-multiple-cluster-subnets-in-layer3-topology.md

```diff
@@ -0,0 +1,273 @@
+
+UDN currently supports only one cluster subnet per IP family in Layer3 topology. This limits IP
+address planning and scalability for large clusters.
+
+## Goals
+
+* Support multiple cluster subnets per IP family in UDN's Layer3 topology.
+* Align UDN’s subnet capabilities with existing support in NAD.
+* Allow dynamic subnet expansion by appending new subnets to an existing UDN.
+* Support only subnets with the same host subnet size(`hostSubnet`).
+
+## Non-Goals
+
+* Supporting multiple subnets in Layer2 or Localnet topologies.
+* Supporting deletion or shrinking of existing subnets.
+
+## Future-Goals
+
+* Support subnets with different host subnet sizes(`hostSubnet`).
+  This feaure requires a OVN commit [27cc274](https://github.com/ovn-org/ovn/commit/27cc274e66acd9e0ed13525f9ea2597804107348).
+
+## Introduction
+
+When using UDN to define a network in Layer3 topology, each node gets a subnet allocated from the
+UDN's `subnets` list per IP family. Currently, UDN limits `subnets` to a single subnet per IP
+family, enforced by CRD validation and full immutability of the UDN spec. This limitation blocks
+practical use cases where operators need to expand IP space incrementally — a feature already
+available through NAD.
+
+This proposal aims to bring UDN's flexibility in line with NAD by supporting multiple cluster
+subnets and enabling controlled mutability.
+
+
+## User-Stories/Use-Cases
+
+### Story 1: Expanding subnet pool for growing clusters
+
+**As a** cluster operator managing a UDN Layer3 network,
+**I want** to add new subnets when existing ones near exhaustion,
+**so that** I can scale the cluster without recreating the UDN object.
+
+
+## Proposed Solution
+
+### API Details
+
+1. Update `UDN`/`CUDN` CRD to allow `layer3.subnets` to accept multiple subnets per IP family.
+2. Allow updates to the `layer3.subnets` field, but only to **append** new entries.
+   - Reject updates that attempt to remove or modify existing subnet entries.
+   - Reject new entries that overlap with existing subnets or use a different `hostSubnet`.
+
+
+### Implementation Details
+
+### Node Subnet Allocation
+
+The current NAD implementation already supports multiple subnets in Layer3 topology. However,
+when a new subnet is added to a NAD, the corresponding network controller is first deleted and then
+recreated with the updated subnet list. This behavior can cause node subnets to be reallocated,
+resulting in unnecessary disruption.
+
+In practice, this issue can often be worked around by restarting the `ovnkube-cluster-manager`, so
+that the controller is recreated with the latest NAD spec and reserve node subnets based on the
+`k8s.ovn.org/node-subnets` annotation.
+
+Ideally, nodes should retain their originally assigned subnets when adding new cluster subnets.
+Changes are required in the `SecondaryNetworkClusterManager` to ensure stable node subnet assignments
+when new subnets are dynamically added.
+
+### OVN Network Topology
```

Comment:
What about the impact on other features? Do we have assumptions from other features that a UDN only contains 1 (or 2 for dual-stack) subnets?

Another element to consider is the impact on the flow tables. How many additional logical OVN constructs are going to be created per-subnet. What does the impact then look like on the flow tables.

Coming back to my earlier point about allowing the mask to be enlarged, do we loose out on options for optimization by allowing non-contiguous CIDRs to be added as subnets.
i.e 10.0.0.0/16 and 10.1.0.0/16 could be summarized as 10.0.0.0/15

Reply:
Since NAD supports multiple subnets per IP family, I don’t think other features assume that UDN contains only one subnet per IP family.

> Another element to consider is the impact on the flow tables.
This shouldn’t be a concern for this OKEP. Whoever uses this feature should evaluate the impact, and if it causes issues, they will most likely need to be addressed on the OVN side.

> Coming back to my earlier point about allowing the mask to be enlarged...
This is not in scope of this OKEP.